File: c/caesar_improved.c
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>


bool is_rel_primes(const int x, const int y)
{
    for (register int i = 2; i < (x > y ? y : x); ++i)
    {
        if (y % i == 0 && x % i == 0)
        {
            return false;
        }
    }

    return true;
}


int* generate_keys(const size_t n)
{
    int* keys = (int*)calloc(2, sizeof(int));

    for (register int i = 1; i < n; ++i)
    {
        for (register int j = 2; j < n; ++j)
        {
            if ((i * j) % n == 1 && is_rel_primes(i, n) && is_rel_primes(j, n))
            {
                keys[0] = i;
                keys[1] = j;
                return keys;
            }
        }
    }

    return NULL;
}


char* encrypt(const char* text, const size_t ke, const size_t n)
{
    const size_t tlength = strlen(text);
    char* result = (char*)malloc(tlength);

    for (register int i = 0; i < tlength; ++i)
    {
        result[i] = (text[i] * ke) % n;
    }

    return result;
}


char* decrypt(const char* text, const size_t kd, const size_t n)
{
    const size_t tlength = strlen(text);
    char* result = (char*)malloc(tlength);

    for (register int i = 0; i < tlength; ++i)
    {
        result[i] = (text[i] * kd) % n;
    }

    return result;
}


int main(int argc, char* argv[])
{
    const int* keys = generate_keys(256);

    if (keys == NULL)
    {
        return -1;
    }

    const int ke = keys[0];
    const int kd = keys[1];

    assert(strcmp(encrypt("cryptography", ke, 256), ")VkP\\M5V#P8k") == 0);
    assert(strcmp(decrypt(")VkP\\M5V#P8k", kd, 256), "cryptography") == 0);

    return 0;
}

File: python/caesar_improved.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
"""
caesar_improved.py: Improved caesar algorithm.
Formulas:
    1) Ek(i) = (i*ke) mod n.
    2) Dk(i) = (i*kd) mod n.

Constraints:
    1) ke & n - relatively prime numbers.
    2) kd & n - relatively prime numbers.
    3) (ke * kd) mod n = 1.
"""


from math import gcd


def generate_keys(n: int = 256) -> tuple:
    """
    generate_keys: Generates keys (ke, kd) for cryptography algorithm.
    """
    for i in range(1, n):
        for j in range(2, n):
            if (i * j) % n == 1 and gcd(i, n) == 1 and gcd(j, n) == 1:
                return i, j


def encrypt(text: str, ke: int, n: int = 256) -> str:
    """
    encrypt: Improved caesar encryption algorithm for ASCII codes.
    """
    return ''.join([chr((ord(x) * ke) % n) for x in text])


def decrypt(text: str, kd: int, n: int = 256) -> str:
    """
    decrypt: Improved caesar decryption algorithm for ASCII codes.
    """
    return ''.join([chr((ord(x) * kd) % n) for x in text])


if __name__ == '__main__':
    ke, kd = generate_keys(256)
    assert encrypt('cryptography', ke) == ')VkP\\M5V#P8k'
    assert decrypt(')VkP\\M5V#P8k', kd) == 'cryptography'
